src/docs/runtime-metadata.md

```diff
@@ -0,0 +1,255 @@
+# Swift Type Metadata
+
+## Background
+
+Swift has a struct which is implicit in the language. It is used for several tasks:
+1 - Determining the classification of the type that is represented (struct, enum, class, tuple, etc)
+2 - Getting access to the type descriptor associated with the type
+3 - Getting access to generic types that specialize the type as well [protocols with associated types](bindings-pats.md)
+4 - Labeling the contents of existential containers representing [protocols](binding-protocols.md)
+5 - Specifying the types of generic parameters to functions
+6 - Providing information used to allocate types
+
+In our case, we will use this type for several things in interoperation with Swift from C#:
+
+1 - Generating existential containers
+2 - Specializing generic types
+3 - Specifying the types of generic parameters in pinvokes into Swift
+4 - Mapping a C# type to an equivalent Swift type and vice versa
+5 - Getting access to [nominal type descriptors](runtime-nominal-type-descriptor.md) to determine size, stride, etc.
+
+Swift type metadata is always a singleton for every unique type in Swift. The Swift runtime enforces this through accessors for getting at the type metadata and it will cache the values. In usage, the type metadata is always passed by reference but they should be considered read-only.
+
+There are two ways to get the type metadata for a Swift type:
+1 - by accessing the type metadata directly by symbolic name (only applies to non-generic nominal types)
+2 - using an accessor
+
+Within accessors, there are two classes of accessor:
+1 - compiler-generated accessors (for all nominal types)
+2 - runtime-provided accessors (for non-nominal types - tuples, closures, etc.)
+
+Generally speaking, it's probably best for us to use the accessors and if we choose, to cache those values as static fields in bound types.
+
+The signature of a type metadata accessor is a request followed by n TypeMetadata objects, one for each generic type:
+```csharp
+public static TypeMetadata Accessor(TypeMetadataRequest request [, TypeMetadata specialization0, TypeMetadata specialization1, ...])
```

Comment:
Do arguments include protocol conformances as well?

Reply:
Protocol conformances are not part of the metadata accessor, but the resulting type metadata with likely (I don't recall off the top of my head) include information about protocol conformance.